File: internal/adapter/sqlite/tx.go
package sqlite

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/jaenster/hoardarr/internal/domain/tx"
)

// maxSnapshotRetries caps the number of times InTx will reopen on a
// retryable busy error. Two flavours surface from modernc.org/sqlite:
//
//   - SQLITE_BUSY_SNAPSHOT (517): a tx opened a read snapshot, then
//     attempted a write after another connection advanced the DB.
//     busy_timeout doesn't help here; only roll back, re-begin, replay.
//
//   - SQLITE_BUSY (5): two connections raced for the write lock and
//     the loser's busy-timeout countdown ran out. With MaxOpenConns >
//     1 and bursty writers (outbox dispatcher + orchestrator + REST
//     mutating endpoints) this still happens occasionally even with a
//     5s timeout; the retry-then-roll-back-and-replay pattern is the
//     same as for 517.
//
// 20 retries with exponential backoff (1ms * 2^attempt, capped at
// 250ms) absorbs the bursty contention pattern we see under CI's
// -race overhead, where multiple tests in the same process are
// hammering the DB concurrently and the writer set turns over fast
// enough that 5 retries can hit 5 consecutive collisions. Worst-case
// total stagger is ~2.5s, well below the 10m go-test timeout and
// the 30s HTTP-client timeout that's the actual outer bound.
const maxSnapshotRetries = 20

// txKey is the unexported context-key type used to attach an active
// *sql.Tx to a context. Repos and the outbox bus extract via TxFromContext
// to participate in the ambient transaction.
type ctxKey struct{}

var txKey ctxKey

// txState carries the *sql.Tx plus a list of post-commit hooks. Hooks
// fire AFTER the outer TX commits successfully — used to defer
// observable side-effects (e.g. waking outbox dispatchers) so they
// only happen when other readers can see the inserted rows. A
// rolled-back TX runs no hooks.
type txState struct {
	tx          *sql.Tx
	commitHooks []func()
}

// TxManager is the SQLite implementation of tx.TransactionManager.
//
// It begins a transaction on InTx entry, attaches the *sql.Tx to the
// context, and commits or rolls back based on fn's return.
//
// Nested InTx calls within the same context join the existing transaction
// (savepoints not used in v0.1).
type TxManager struct {
	db *DB
}

// Compile-time check that TxManager satisfies the domain port.
var _ tx.TransactionManager = (*TxManager)(nil)

// NewTxManager returns a TransactionManager backed by db.
func NewTxManager(db *DB) *TxManager {
	return &TxManager{db: db}
}

// InTx runs fn inside a transaction.
//
// If ctx already carries a *sql.Tx, fn is called with the same ctx and
// the existing transaction is reused (joined). Errors from joined fn
// calls propagate up; the outer InTx still controls commit/rollback.
//
// If ctx has no transaction, a fresh one is begun. fn returning nil
// commits; non-nil rollbacks. Panics in fn rollback and re-panic.
func (m *TxManager) InTx(ctx context.Context, fn func(ctx context.Context) error) error {
	if existing := TxFromContext(ctx); existing != nil {
		return fn(ctx)
	}
	var lastErr error
	for attempt := 0; attempt < maxSnapshotRetries; attempt++ {
		err := m.runTx(ctx, fn)
		if err == nil {
			return nil
		}
		if !isRetryableBusy(err) {
			return err
		}
		lastErr = err
		// Exponential backoff with a 250ms cap. Spinning every
		// attempt+1 ms (the previous linear scheme) doesn't give the
		// other writer enough room to finish on a slow runner.
		delay := time.Duration(1<<attempt) * time.Millisecond
		if delay > 250*time.Millisecond {
			delay = 250 * time.Millisecond
		}
		select {
		case <-time.After(delay):
		case <-ctx.Done():
			return ctx.Err()
		}
	}
	return fmt.Errorf("intx: busy after %d retries: %w", maxSnapshotRetries, lastErr)
}

// runTx is one transaction attempt. Separated so InTx can replay on
// SQLITE_BUSY_SNAPSHOT / SQLITE_BUSY.
func (m *TxManager) runTx(ctx context.Context, fn func(ctx context.Context) error) error {
	sqlTx, err := m.db.BeginTx(ctx, nil)
	if err != nil {
		return fmt.Errorf("begin tx: %w", err)
	}
	state := &txState{tx: sqlTx}
	defer func() {
		if p := recover(); p != nil {
			_ = sqlTx.Rollback()
			panic(p)
		}
	}()
	if err := fn(context.WithValue(ctx, txKey, state)); err != nil {
		if rbErr := sqlTx.Rollback(); rbErr != nil && !errors.Is(rbErr, sql.ErrTxDone) {
			return fmt.Errorf("rollback after %v: %w", err, rbErr)
		}
		return err
	}
	if err := sqlTx.Commit(); err != nil {
		return fmt.Errorf("commit: %w", err)
	}
	// Fire post-commit hooks. We deliberately run them after a
	// successful commit so observers (e.g. outbox dispatchers being
	// nudged) see the just-inserted rows when they query.
	for _, hook := range state.commitHooks {
		hook()
	}
	return nil
}

// isRetryableBusy matches the two flavours of busy / lock-contention
// modernc.org/sqlite surfaces: plain SQLITE_BUSY (5) and the snapshot-
// conflict variant SQLITE_BUSY_SNAPSHOT (517). Both clear with a tx
// replay; neither benefits from blindly waiting (517 never times out,
// and a 5 that already exhausted busy_timeout won't relax on its own).
func isRetryableBusy(err error) bool {
	if err == nil {
		return false
	}
	s := err.Error()
	// 517 — snapshot conflict.
	if strings.Contains(s, "SQLITE_BUSY_SNAPSHOT") ||
		strings.Contains(s, "(517)") {
		return true
	}
	// 5 — plain SQLITE_BUSY (write-write race after busy_timeout).
	if strings.Contains(s, "database is locked (5)") ||
		strings.Contains(s, "SQLITE_BUSY") {
		return true
	}
	return false
}

// TxFromContext returns the *sql.Tx attached to ctx, or nil if none.
//
// Used by repository implementations and the outbox bus to bind their
// queries to the ambient transaction. If nil, the caller should fall
// back to db-level Exec/Query (single-statement, auto-committed).
func TxFromContext(ctx context.Context) *sql.Tx {
	if st, ok := ctx.Value(txKey).(*txState); ok && st != nil {
		return st.tx
	}
	return nil
}

// OnTxCommit registers fn to run after the ambient transaction
// commits successfully. If ctx carries no transaction, fn runs
// immediately. fn does not run if the transaction is rolled back.
//
// Used for side-effects that observers (other goroutines) must not
// see before the TX is durable — most notably the outbox bus waking
// up dispatcher goroutines, which would otherwise SELECT and find
// nothing while the INSERT is still pending in an uncommitted TX.
func OnTxCommit(ctx context.Context, fn func()) {
	if st, ok := ctx.Value(txKey).(*txState); ok && st != nil {
		st.commitHooks = append(st.commitHooks, fn)
		return
	}
	fn()
}

// ExecContext runs the query inside the ambient transaction if one is
// in ctx, otherwise directly on the DB. Repositories should call this
// instead of *sql.DB.ExecContext to be tx-aware.
func (db *DB) ExecCtx(ctx context.Context, query string, args ...any) (sql.Result, error) {
	if t := TxFromContext(ctx); t != nil {
		return t.ExecContext(ctx, query, args...)
	}
	return db.ExecContext(ctx, query, args...)
}

// QueryCtx runs the query inside the ambient tx if present.
func (db *DB) QueryCtx(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
	if t := TxFromContext(ctx); t != nil {
		return t.QueryContext(ctx, query, args...)
	}
	return db.QueryContext(ctx, query, args...)
}

// QueryRowCtx runs the single-row query inside the ambient tx if present.
func (db *DB) QueryRowCtx(ctx context.Context, query string, args ...any) *sql.Row {
	if t := TxFromContext(ctx); t != nil {
		return t.QueryRowContext(ctx, query, args...)
	}
	return db.QueryRowContext(ctx, query, args...)
}

File: src/store/tx.zig
//! Transactions: `BEGIN IMMEDIATE`, nesting-by-joining, post-commit
//! hooks, and replay on lock contention.
//!
//! ## Why the connection carries the transaction
//!
//! The Go build threaded the active `*sql.Tx` through a
//! `context.Context` so a repository method could join whatever
//! transaction its caller had open. Zig has no ambient context, and it
//! does not need one: a connection is owned by a single thread
//! (`sqlite.zig`), so "the transaction open on this connection" *is*
//! the ambient transaction. `Conn.tx_depth` holds it, and a repository
//! that writes through the connection automatically participates.
//!
//! ## Nesting
//!
//! An inner `inTx` joins the outer one — it runs the body with no SQL
//! of its own, and the outermost frame decides commit or rollback.
//! There are no savepoints: a nested rollback in this system always
//! means "abandon the whole unit of work", and partial rollback would
//! let a caller commit half an aggregate.
//!
//! ## Post-commit hooks
//!
//! `onCommit` defers a side-effect until after `COMMIT` returns. This
//! is not a convenience — it is a correctness fix. The outbox bus
//! nudges its dispatcher threads when it publishes an event, and until
//! the outer transaction commits those `INSERT`s are invisible to every
//! other connection. A dispatcher woken too early queries, sees
//! nothing, and goes back to sleep until its next poll tick — which at
//! the 5-second production interval turned a sub-millisecond handoff
//! into a five-second stall, and tripped the end-to-end suite about two
//! runs in three. Hooks run only on success; a rolled-back transaction
//! runs none.
//!
//! ## Replay
//!
//! Two result codes mean "the transaction did not happen, try again":
//!
//!   * `SQLITE_BUSY` (5) — lost the race for the write lock after
//!     `busy_timeout` expired.
//!   * `SQLITE_BUSY_SNAPSHOT` (517) — a read snapshot went stale under
//!     us. `busy_timeout` never clears this one; only rolling back and
//!     replaying does.
//!
//! `BEGIN IMMEDIATE` makes the second case rare (the write lock is
//! claimed before any read), but "rare" is not "never" when several
//! writer threads share one file, so `inTx` replays either way.

const std = @import("std");
const sqlite = @import("sqlite.zig");
const sys = @import("../posix/sys.zig");

const Conn = sqlite.Conn;

/// Cap on replays before `inTx` gives up with the last error.
///
/// Twenty attempts with the backoff below stagger over ~2.5 s in the
/// worst case. That absorbs the bursty pattern seen when the
/// orchestrator's drainer, the outbox dispatchers and a handful of
/// mutating API requests all want the write lock inside the same few
/// milliseconds, and it stays far below the outer bound that actually
/// matters — the HTTP client's timeout.
pub const max_retries: usize = 20;

const backoff_base_ns: u64 = std.time.ns_per_ms;
const backoff_cap_ns: u64 = 250 * std.time.ns_per_ms;

/// Open a transaction, or join the one already open on `conn`.
///
/// `BEGIN IMMEDIATE`, never a bare `BEGIN`: the default deferred
/// behaviour opens a read snapshot and only upgrades to a write lock on
/// the first mutation, at which point a concurrent commit surfaces as
/// `SQLITE_BUSY_SNAPSHOT` — a code `busy_timeout` cannot wait out.
/// Claiming the lock upfront turns writer contention back into the
/// bounded wait that `busy_timeout` is designed for.
pub fn begin(conn: *Conn) sqlite.Error!void {
    if (conn.tx_depth > 0) {
        conn.tx_depth += 1;
        return;
    }
    // `execute` rather than `exec`: it goes through the statement cache, so
    // opening a transaction costs a `sqlite3_reset` instead of a parse plus
    // a NUL-terminated copy of the SQL. The orchestrator's drainer opens
    // one of these every few hundred milliseconds and every repository
    // write is wrapped in one, so this is a hot path.
    try conn.execute("BEGIN IMMEDIATE", .{});
    conn.tx_depth = 1;
}

/// Close the current nesting level. The outermost level commits and
/// then runs the queued post-commit hooks.
pub fn commit(conn: *Conn) sqlite.Error!void {
    std.debug.assert(conn.tx_depth > 0);
    if (conn.tx_depth > 1) {
        conn.tx_depth -= 1;
        return;
    }
    try conn.execute("COMMIT", .{});
    conn.tx_depth = 0;
    runHooks(conn);
}

/// Abandon the whole transaction regardless of nesting depth, and drop
/// every queued hook unrun.
///
/// Errors are swallowed deliberately: this is the unwind path, the
/// caller already has the error that brought it here, and a failing
/// `ROLLBACK` means the transaction is gone anyway.
pub fn rollback(conn: *Conn) void {
    if (conn.tx_depth == 0) return;
    conn.tx_depth = 0;
    conn.execute("ROLLBACK", .{}) catch {};
    conn.commit_hooks.clearRetainingCapacity();
}

/// Register `run(ctx)` to fire after the outermost transaction commits.
/// With no transaction open it fires immediately, which keeps callers
/// from having to care whether they are inside one.
pub fn onCommit(
    conn: *Conn,
    ctx: ?*anyopaque,
    run: *const fn (?*anyopaque) void,
) std.mem.Allocator.Error!void {
    if (conn.tx_depth == 0) {
        run(ctx);
        return;
    }
    try conn.commit_hooks.append(conn.gpa, .{ .ctx = ctx, .run = run });
}

fn runHooks(conn: *Conn) void {
    // Snapshot the length first: a hook may open its own transaction and
    // register more hooks, and those belong to that transaction, not to
    // the one just committed.
    const n = conn.commit_hooks.items.len;
    for (conn.commit_hooks.items[0..n]) |hook| hook.run(hook.ctx);
    conn.commit_hooks.replaceRange(conn.gpa, 0, n, &.{}) catch {
        conn.commit_hooks.clearRetainingCapacity();
    };
}

/// Run `body(ctx, conn)` inside a transaction, replaying it on lock
/// contention.
///
/// `body` returning an error rolls back and the error propagates
/// unchanged — a domain error must not be mistaken for a storage
/// failure. When `conn` already has a transaction open, `body` simply
/// joins it: no replay (the outer frame owns the outcome), and the error
/// propagates so the outer frame can roll back.
///
/// The error set is inferred because it is the union of `sqlite.Error`
/// and whatever `body` returns.
pub fn inTx(conn: *Conn, ctx: anytype, comptime body: anytype) !void {
    if (conn.tx_depth > 0) {
        // Joined: count the level so a nested commit is symmetric with
        // its begin, but let the outer frame decide the outcome.
        try begin(conn);
        errdefer rollback(conn);
        try body(ctx, conn);
        try commit(conn);
        return;
    }

    var attempt: usize = 0;
    while (true) : (attempt += 1) {
        const failure = attempt: {
            begin(conn) catch |e| break :attempt e;
            body(ctx, conn) catch |e| {
                rollback(conn);
                break :attempt e;
            };
            commit(conn) catch |e| {
                rollback(conn);
                break :attempt e;
            };
            return;
        };

        // Only storage-level contention is replayable. Anything else —
        // a constraint violation, an I/O error, a domain error from the
        // body — is returned as-is.
        const retryable = switch (@as(anyerror, failure)) {
            error.Busy, error.BusySnapshot, error.Locked => true,
            else => false,
        };
        if (!retryable or attempt + 1 >= max_retries) return failure;

        // Exponential backoff, capped. A fixed 1 ms spin (the shape this
        // started as) does not give the other writer room to finish on a
        // loaded machine, and the collision then just repeats.
        const shift: u6 = @intCast(@min(attempt, 8));
        sys.sleep(@min(backoff_base_ns << shift, backoff_cap_ns));
    }
}

/// True when `conn` has a transaction open. The Zig equivalent of Go's
/// `TxFromContext(ctx) != nil`.
pub fn inTransaction(conn: *const Conn) bool {
    return conn.tx_depth > 0;
}

// ---------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------

const t = std.testing;
const support = sqlite.testing_support;

fn scratch() !*Conn {
    const conn = try support.openMemory();
    try conn.exec("CREATE TABLE t(id INTEGER PRIMARY KEY, v TEXT NOT NULL)");
    return conn;
}

fn insertOne(_: void, conn: *Conn) !void {
    try conn.execute("INSERT INTO t(v) VALUES (?)", .{"x"});
}

test "a committed transaction persists its writes" {
    const conn = try scratch();
    defer conn.close();

    try inTx(conn, {}, insertOne);
    try t.expectEqual(@as(i64, 1), try conn.scalarInt("SELECT COUNT(*) FROM t", .{}));
    try t.expect(!inTransaction(conn));
}

test "a body error rolls back and propagates unchanged" {
    const conn = try scratch();
    defer conn.close();

    const failing = struct {
        fn run(_: void, cn: *Conn) !void {
            try cn.execute("INSERT INTO t(v) VALUES (?)", .{"x"});
            return error.CallerChangedTheirMind;
        }
    }.run;

    try t.expectError(error.CallerChangedTheirMind, inTx(conn, {}, failing));
    try t.expectEqual(@as(i64, 0), try conn.scalarInt("SELECT COUNT(*) FROM t", .{}));
    try t.expect(!inTransaction(conn));
}

test "a nested transaction joins the outer one" {
    const conn = try scratch();
    defer conn.close();

    const outer = struct {
        fn run(_: void, cn: *Conn) !void {
            try t.expect(inTransaction(cn));
            try inTx(cn, {}, insertOne);
            // Still inside the outer transaction after the inner one
            // "commits" — the inner commit only decremented the depth.
            try t.expect(inTransaction(cn));
        }
    }.run;

    try inTx(conn, {}, outer);
    try t.expectEqual(@as(i64, 1), try conn.scalarInt("SELECT COUNT(*) FROM t", .{}));
}

test "an inner error rolls the whole outer transaction back" {
    const conn = try scratch();
    defer conn.close();

    const nest = struct {
        fn inner(_: void, cn: *Conn) !void {
            try cn.execute("INSERT INTO t(v) VALUES (?)", .{"inner"});
            return error.Nope;
        }
        fn outer(_: void, cn: *Conn) !void {
            try cn.execute("INSERT INTO t(v) VALUES (?)", .{"outer"});
            try inTx(cn, {}, inner);
        }
    };

    try t.expectError(error.Nope, inTx(conn, {}, nest.outer));
    try t.expectEqual(@as(i64, 0), try conn.scalarInt("SELECT COUNT(*) FROM t", .{}));
    try t.expect(!inTransaction(conn));
}

test "begin uses IMMEDIATE so the write lock is claimed upfront" {
    // Two connections onto one file. The first claims the write lock in
    // BEGIN itself, so the second's BEGIN fails with BUSY *before* it
    // has read anything — which is the whole point of IMMEDIATE over the
    // deferred default, where the failure would instead land on the
    // first write as BUSY_SNAPSHOT.
    const path = try support.tempPath(t.allocator);
    defer t.allocator.free(path);
    defer support.removeTempFile(t.allocator, path);

    const a = try sqlite.Conn.open(t.allocator, path, .{ .busy_timeout_ms = 0 });
    defer a.close();
    try a.exec("CREATE TABLE t(id INTEGER PRIMARY KEY, v TEXT)");

    const b = try sqlite.Conn.open(t.allocator, path, .{ .busy_timeout_ms = 0 });
    defer b.close();

    try begin(a);
    try a.execute("INSERT INTO t(v) VALUES (?)", .{"first"});
    try t.expectError(error.Busy, begin(b));
    try commit(a);

    // Lock released: b can now take it.
    try begin(b);
    try commit(b);
}

test "inTx gives up after max_retries against a permanently held lock" {
    const path = try support.tempPath(t.allocator);
    defer t.allocator.free(path);
    defer support.removeTempFile(t.allocator, path);

    const a = try sqlite.Conn.open(t.allocator, path, .{ .busy_timeout_ms = 0 });
    defer a.close();
    try a.exec("CREATE TABLE t(id INTEGER PRIMARY KEY, v TEXT NOT NULL)");

    const b = try sqlite.Conn.open(t.allocator, path, .{ .busy_timeout_ms = 0 });
    defer b.close();

    try begin(a);
    try a.execute("INSERT INTO t(v) VALUES (?)", .{"held"});

    // The assertion is that it terminates with the retryable error
    // rather than hanging. The backoff sums to ~2.5 s worst case, which
    // is real but bounded.
    try t.expectError(error.Busy, inTx(b, {}, insertOne));
    try t.expect(!inTransaction(b));
    try commit(a);
}

test "post-commit hooks run after COMMIT, in registration order" {
    const conn = try scratch();
    defer conn.close();

    const Observer = struct {
        conn: *Conn,
        order: [3]u8 = .{ 0, 0, 0 },
        n: usize = 0,
        rows_when_first_ran: i64 = -1,

        fn note(self: *@This(), tag: u8) void {
            if (self.n == 0) {
                // The row must already be visible: that is the property
                // the whole hook mechanism exists to guarantee.
                self.rows_when_first_ran =
                    self.conn.scalarInt("SELECT COUNT(*) FROM t", .{}) catch -1;
            }
            self.order[self.n] = tag;
            self.n += 1;
        }
        fn a(p: ?*anyopaque) void {
            @as(*@This(), @ptrCast(@alignCast(p.?))).note('a');
        }
        fn b(p: ?*anyopaque) void {
            @as(*@This(), @ptrCast(@alignCast(p.?))).note('b');
        }
        fn c(p: ?*anyopaque) void {
            @as(*@This(), @ptrCast(@alignCast(p.?))).note('c');
        }
    };
    var obs = Observer{ .conn = conn };

    const body = struct {
        fn run(o: *Observer, cn: *Conn) !void {
            try cn.execute("INSERT INTO t(v) VALUES (?)", .{"x"});
            try onCommit(cn, o, Observer.a);
            try onCommit(cn, o, Observer.b);
            try onCommit(cn, o, Observer.c);
            // Nothing has fired yet — the rows are still invisible.
            try t.expectEqual(@as(usize, 0), o.n);
        }
    }.run;

    try inTx(conn, &obs, body);
    try t.expectEqual(@as(usize, 3), obs.n);
    try t.expectEqualSlices(u8, "abc", &obs.order);
    try t.expectEqual(@as(i64, 1), obs.rows_when_first_ran);
}

test "a rolled-back transaction runs no hooks" {
    const conn = try scratch();
    defer conn.close();

    const Flag = struct {
        fired: bool = false,
        fn set(p: ?*anyopaque) void {
            @as(*@This(), @ptrCast(@alignCast(p.?))).fired = true;
        }
    };
    var flag = Flag{};

    const body = struct {
        fn run(f: *Flag, cn: *Conn) !void {
            try onCommit(cn, f, Flag.set);
            return error.Abandon;
        }
    }.run;

    try t.expectError(error.Abandon, inTx(conn, &flag, body));
    try t.expect(!flag.fired);
    try t.expectEqual(@as(usize, 0), conn.commit_hooks.items.len);
}

test "onCommit outside a transaction runs immediately" {
    const conn = try scratch();
    defer conn.close();

    const Flag = struct {
        fired: bool = false,
        fn set(p: ?*anyopaque) void {
            @as(*@This(), @ptrCast(@alignCast(p.?))).fired = true;
        }
    };
    var flag = Flag{};
    try onCommit(conn, &flag, Flag.set);
    try t.expect(flag.fired);
}

test "hooks registered by a hook belong to their own transaction" {
    const conn = try scratch();
    defer conn.close();

    const Nested = struct {
        conn: *Conn,
        outer_ran: bool = false,
        inner_ran: bool = false,

        fn outerHook(p: ?*anyopaque) void {
            const self: *@This() = @ptrCast(@alignCast(p.?));
            self.outer_ran = true;
            // No transaction is open by now, so this runs immediately
            // rather than being left dangling in the outer queue.
            onCommit(self.conn, p, innerHook) catch {};
        }
        fn innerHook(p: ?*anyopaque) void {
            @as(*@This(), @ptrCast(@alignCast(p.?))).inner_ran = true;
        }
    };
    var n = Nested{ .conn = conn };

    const body = struct {
        fn run(s: *Nested, cn: *Conn) !void {
            try onCommit(cn, s, Nested.outerHook);
        }
    }.run;

    try inTx(conn, &n, body);
    try t.expect(n.outer_ran);
    try t.expect(n.inner_ran);
    try t.expectEqual(@as(usize, 0), conn.commit_hooks.items.len);
}

test "a transaction is invisible to another connection until it commits" {
    const path = try support.tempPath(t.allocator);
    defer t.allocator.free(path);
    defer support.removeTempFile(t.allocator, path);

    const w = try sqlite.Conn.open(t.allocator, path, .{});
    defer w.close();
    try w.exec("CREATE TABLE t(id INTEGER PRIMARY KEY, v TEXT)");

    const r = try sqlite.Conn.open(t.allocator, path, .{});
    defer r.close();

    try begin(w);
    try w.execute("INSERT INTO t(v) VALUES (?)", .{"pending"});
    // WAL: the reader keeps its own snapshot and sees nothing.
    try t.expectEqual(@as(i64, 0), try r.scalarInt("SELECT COUNT(*) FROM t", .{}));
    try commit(w);
    try t.expectEqual(@as(i64, 1), try r.scalarInt("SELECT COUNT(*) FROM t", .{}));
}
